File: advancedGraphs/alienDictionary.py
class Solution:
    def foreignDictionary(self, words: list[str]) -> str:
        adj = {c: [] for w in words for c in w}

        for i in range(len(words) - 1):
            w1, w2 = words[i], words[i + 1]
            minLen = min(len(w1), len(w2))
            if (len(w1) > len(w2)) and (w1[:minLen] == w2[:minLen]):
                return ""
            for j in range(minLen):
                if w1[j] != w2[j]:
                    adj[w1[j]].append(w2[j])
                    break
        visit = {} # False = visited, True = current path
        res = []

        def dfs(c):
            if c in visit:
                return visit[c]
            
            visit[c] = True
            for nei in adj[c]:
                if dfs(nei): 
                    return True
            visit[c] = False
            res.append(c)

        for c in adj:
            if dfs(c):
                return ""
        res.reverse()
        return "".join(res)


if __name__ == '__main__':
    Input = ["hrn","hrf","er","enn","rfnn"]
    print('Input : {}'.format(Input))
    print('Solution : {}'.format(Solution().foreignDictionary(Input)))

File: advancedGraphs/cppRealizations/alienDictionary.cpp
#include <vector>
#include <iostream>
#include <map>
#include <algorithm>

using namespace std;


class Solution {
public:
    string foreignDictionary(vector<string>& words) {
        map<char, vector<char>> adj;
        for(auto word : words){
            for(auto c : word){
                adj[c] = vector<char>();
            }
        }

        for(int i = 0; i < words.size() - 1; i++){
            string word1 = words[i];
            string word2 = words[i + 1];
            int len = min(word1.size(), word2.size());
            if(word1.size() > word2.size() && word1.substr(0, len) == word2){
                return "";
            }
            for(int j = 0; j < len; j++){
                if(word1[j] != word2[j]){
                    adj[word1[j]].push_back(word2[j]);
                    break;
                }
            }
        }
        map<char, bool> visit;
        vector<char> result;

        for(auto c : adj){
            if(dfs(c.first, adj, visit, result)){
                return "";
            }
        }
        reverse(result.begin(), result.end());
        string res = "";
        for(auto c : result){
            res += c;
        }
        return res;
    }
private:
    bool dfs(char c, map<char, vector<char>>& adj, map<char, bool>& visit, vector<char>& result){
        if(visit.find(c) != visit.end()){
            return visit[c];
        }

        visit[c] = true;
        for(char nei : adj[c]){
            if(dfs(nei, adj, visit, result)){
                return true;
            }
        }
        visit[c] = false;
        result.push_back(c);
        return false;
    }
};


ostream& operator<<(ostream& out, const vector<string>& str){
    out << "[";
    for(size_t i = 0; i < str.size(); i++){
        if(i < str.size() - 1){
            out << str[i] << ",";
        }else{
            out << str[i];
        }
    }
    out << "]";
    return out;
}


int main(){
    vector<string> input = {"hrn","hrf","er","enn","rfnn"};
    cout << "Input: " << input << endl;
    cout << "Solution: " << Solution().foreignDictionary(input) << endl;
    return 0;
}